### ISSUE TITLE
Support multiple IPv4 address ranges for service Cluster IPs

### ISSUE THREAD

**Issue by PaulFurtado** [kubernetes/kubernetes#104088]
#### What would you like to be added:
Kubernetes should allow you to add additional IP ranges for ClusterIP services as your clusters grow.

#### Why is this needed:
Kubernetes clusters naturally grow over time and at large enterprises you can't hand out massive address ranges to each kubernetes cluster on the day they are created given how small the available number of private IPv4 IPs is.

We currently have a cluster which has existed since the beginning of our kubernetes endeavors several years ago and has utilized all possible ClusterIPs. The adjacent IP ranges are utilized, so we can't simply restart the control plane with a different prefixlen on the CIDR. So right now, the only way to remediate this situation is to incur significant downtime by:
1. Deleting all ClusterIP services in the cluster
2. Restarting the control plane and kube-proxies with a new range
3. Recreating all ClusterIP services so that they'll have IPs assigned from the new range
4. Reconfiguring and/or restarting all systems that relied on the fact that ClusterIPs are supposed to be static. This effectively means any pod that uses incluster config to access the API, any systems that are bad at re-resolving DNS and communicate with ClusterIP services, and so much more in various applications.

If the kubernetes control-plane could simply allow multiple IPv4 ranges as an argument then all that would be required is restarting several kubernetes components and moving on, with no downtime incurred for existing services. The caveat is that if your cluster configuration requires that kube-proxy or your CNI plugins need to know the ClusterIP range to function, then there's a risk that new ClusterIPs may not be routeable while the changes make their way out to all nodes, however, that constraint feels so much more acceptable than the alternative of breaking every service and it can be well-documented that you must roll the change out to nodes and potentially must restart all pods before updating the control-plan to ensure this is seamless.

**Comment by aojea** [comment]
we have discussed this heavily in this thread https://github.com/kubernetes/enhancements/pull/1881 ,
 TL;DR we may be there incrementally, but don't expect this to be a thing soon, this is complex to implement, there are a lot of internal details that needs to be solved first

**Comment by PaulFurtado** [comment]
I'll also mention here that I worked around this in our infrastructure by shifting the IP allocation to an admission controller and widening the range the API server was told to use. So for example, in a test cluster, the IP range was a tiny `172.19.32.0/22` originally. So I told the API server that the range was now `172.16.0.0/12`, and then created an admission controller that allocates IPs out of only  `172.19.32.0/22`, `172.19.36.0/24`, and `172.19.44.0/22`, with special handling to ensure the `kubernetes.default` service always is the first IP of the first range in the list, even if something deleted it. I just had to delete `/registry/ranges/serviceips` in etcd to get the apiservers to recreate it and cooperate.

This also did require modifying our in-house ipvlan-based CNI plugin to sync some updated routes into all pod namespaces, so it remains that YMMV based on your CNI plugin, but this went out really smoothly for us with no downtime.

**Comment by aojea** [comment]
> I'll also mention here that I worked around this in our infrastructure by shifting the IP allocation to an admission controller and widening the range the API server was told to use.

Does it handle the case when the apiserver fails to create the service after the admission controller allocated the IP?

create Svc -> admission Allocate IP X --> apiserver (fails to allocate, create service, or simply dies) --> ??? 

how admission controller knows the IP wasn't allocated correctly?

**Comment by aojea** [comment]
/reopen
/assign @aojea 

Will be fixed by https://github.com/kubernetes/enhancements/issues/1880

**Comment by aojea** [comment]
This is beta disabled by default now https://github.com/kubernetes/kubernetes/pull/125021

**Comment by aojea** [comment]
/close

This will be GA in 1.33 https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs
